bpf/lib/nodeport.h

```diff
@@ -1005,31 +1128,41 @@ nodeport_rev_dnat_fwd_ipv6(struct __ctx_buff *ctx, struct trace_ctx *trace)
 	if (!revalidate_data(ctx, &data, &data_end, &ip6))
 		return DROP_INVALID;
 
-	ret = lb6_extract_tuple(ctx, ip6, &l4_off, &tuple);
+	ret = lb6_extract_tuple(ctx, ip6, ETH_HLEN, &l4_off, &tuple);
 	if (ret < 0) {
 		if (ret == DROP_NO_SERVICE || ret == DROP_UNKNOWN_L4)
 			return CTX_ACT_OK;
 		return ret;
 	}
 
-	if (!ct_has_nodeport_egress_entry6(get_ct_map6(&tuple), &tuple))
+	if (!ct_has_nodeport_egress_entry6(get_ct_map6(&tuple), &tuple,
+					   is_defined(ENABLE_DSR)))
 		return CTX_ACT_OK;
 
 	ret = ct_lb_lookup6(get_ct_map6(&tuple), &tuple, ctx, l4_off, CT_INGRESS,
 			    &ct_state, &trace->monitor);
 
-	if (ret == CT_REPLY && ct_state.node_port && ct_state.rev_nat_index) {
-		struct csum_offset csum_off = {};
-
+	if (ret == CT_REPLY) {
 		trace->reason = TRACE_REASON_CT_REPLY;
-		csum_l4_offset_and_flags(tuple.nexthdr, &csum_off);
 
-		ret = lb6_rev_nat(ctx, l4_off, &csum_off, ct_state.rev_nat_index,
-				  &tuple, REV_NAT_F_TUPLE_SADDR);
-		if (IS_ERR(ret))
-			return ret;
+		if (ct_state.node_port && ct_state.rev_nat_index) {
+			struct csum_offset csum_off = {};
 
-		ctx_snat_done_set(ctx);
+			csum_l4_offset_and_flags(tuple.nexthdr, &csum_off);
+
+			ret = lb6_rev_nat(ctx, l4_off, &csum_off, ct_state.rev_nat_index,
+					  &tuple, REV_NAT_F_TUPLE_SADDR);
+			if (IS_ERR(ret))
+				return ret;
+
+			ctx_snat_done_set(ctx);
```

Comment:
Nit for a follow up PR - I think s/ctx_snat_done_set/ctx_snat_not_needed/ would be better naming. In the case of DSR we don't do any SNAT (we  do rev-DNAT).

Reply:
Yep, that's a good follow-on cleanup :+1:. It's not obvious that we've been overloading the flag with other semantics.